source/server.go

```diff
@@ -105,14 +113,20 @@ func (s *Server) recvRecords(stream pb.SourceService_StreamServer) error {
 	}
 }
 
-func (s *Server) SendAck(position sdk.Position) error {
+func (s *Server) SendAck(ctx context.Context, position sdk.Position) error {
 	stream := s.stream.Load()
-	if stream == nil {
-		return fmt.Errorf("no stream is open")
+	for stream == nil {
+		select {
+		case <-s.streamIsSetCh:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+		stream = s.stream.Load()
 	}
 	err := (*stream).Send(&pb.Ack{AckPosition: position})
 	if err != nil {
-		return fmt.Errorf("error while sending ack into stream: %w", err)
+		sdk.Logger(ctx).Err(err).Msg("error while sending ack into stream")
+		return s.SendAck(ctx, position)
```

Comment:
Btw, we should probably have a limit on how many times we retry. Imagine that we try to send an ack and get an error, but the reason is something other than a closed stream. When we retry the stream will be open so we will immediately try to send the ack again, and we will end up in an endless loop. I'm not sure which errors exactly are possible here, but I imagine that different things can happen here and I'd feel better if we stay on the safe side and stop retrying after let's say 5 tries.